### PR TITLE
1365 1377 Adding Process Exploration

### PR DESCRIPTION
**Pull request by cpholguera**
This PR covers issues #1365 and #1377.

### CODE REVIEW

**Comment by OMTGreleaser** [comment]
Results for commit 9fc5ac2abb56a6f3655a8efb2509fcdd14a3843c: Broken link result: [!] Error(s) found by the Links validator: 8 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

**Comment by OMTGreleaser** [comment]
Results for commit 570ff2b5d0e44fe7612e1c8b31280581576511f9: Broken link result: [!] Error(s) found by the Links validator: 4 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

**Comment by OMTGreleaser** [comment]
Results for commit 87cd80837a250368f8663661e8842980818c8757: Broken link result: [!] Error(s) found by the Links validator: 5 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

File: Document/0x05c-Reverse-Engineering-and-Tampering.md [hunk 1]
```diff
@@ -1497,7 +1488,368 @@ The hooked function outputted the decrypted string. You extracted the secret str
...
+- After doing a search, obtaining the location of a certain offset in the memory map.
+- Performing a memory dump and inspect or reverse engineer the binary data _offline_.
+- Reverse engineering a native library while it's running.
+
+As you can see, these tasks are rather supportive and/or passive, they'll help us collect data and information that will support other techniques. Therefore, they're normally used in combination with other techniques such as method hooking.
+
+In the following sections you will be using r2frida to retrieve information straight from the app runtime. First start by opening an r2frida session to the target app (e.g. [HelloWorld JNI](https://github.com/OWASP/owasp-mstg/raw/master/Samples/Android/01_HelloWorld-JNI/HelloWord-JNI.apk "HelloWorld JNI") APK) that should be running on your Android phone (connected per USB). Use the following command:
+
+```bash
+$ r2 frida://usb//sg.vantagepoint.helloworldjni
```
Comment (commjoen) [inline Document/0x05c-Reverse-Engineering-and-Tampering.md]: where do we explain on how to get r2 frida working? (e.g. a link to the readme for installation instructions)

File: Document/0x05c-Reverse-Engineering-and-Tampering.md [hunk 2]
```diff
@@ -1497,7 +1488,368 @@ The hooked function outputted the decrypted string. You extracted the secret str
...
+0x0000007dc065d000 linux-vdso.so.1
+0x0000007dc065f000 linker64
+```
+
+As you might expect you can correlate the addresses of the libraries with the memory maps: e.g. the native library of the app is located at `0x0000007d1c499000` and optimized dex (base.odex) at `0x0000007d10dd0000`.
+
+You can also use objection to display the same information.
+
+```shell
+sg.vantagepoint.helloworldjni on (google: 8.1.0) [usb] # memory list modules
```
Comment (commjoen) [inline Document/0x05c-Reverse-Engineering-and-Tampering.md]: i am missing the command to get here?

File: Document/0x05c-Reverse-Engineering-and-Tampering.md [hunk 3]
```diff
@@ -1497,7 +1488,368 @@ The hooked function outputted the decrypted string. You extracted the secret str
...
+0x0000000013100000 - 0x0000000013140000 rw- /dev/ashmem/dalvik-main space (region space) (deleted)
+0x0000007d30a00000 - 0x0000007d30c00000 rw-
+0x0000007d30a00000 - 0x0000007d30c00000 rw-
+0x0000007d30a00000 - 0x0000007d30c00000 rw-
+0x0000007d30a00000 - 0x0000007d30c00000 rw-
+```
+
+They are in the same rw- region as one of the previous strings (`0x0000007d30a00000`). Note that searching for the wide versions of strings is sometimes the only way to find them as you'll see in the following section.
+
+In-memory search can be very useful to quickly know if certain data is located in the main app binary, inside a shared library or in another region. You may also use it to test the behavior of the app regarding how the data is kept in memory. For instance, you could analyze an app that performs a login and search for occurrences of the user password. Also, you may check if you still can find the password in memory after the login is completed to verify if this sensitive data is wiped from memory after its use.
```
Comment (commjoen) [inline Document/0x05c-Reverse-Engineering-and-Tampering.md]: Maybe this can be extended to explain that using keys to encrypt data in a binary can be found like this as well? (to further push people towards using the AndroidKeyStore ^^)

File: Document/0x05c-Reverse-Engineering-and-Tampering.md [hunk 4]
```diff
@@ -1497,7 +1488,368 @@ The hooked function outputted the decrypted string. You extracted the secret str
...
+
+###### Memory Dump
+
+Wether you are using a rooted phone with frida-server installed or a non-rooted device, you can dump the app's process memory with [objection](https://github.com/sensepost/objection "Objection") and [Fridump](https://github.com/Nightbringer21/fridump "Fridump"). To take advantage of these tools on a non-rooted device, the Android app must be repackaged with `frida-gadget.so` and re-signed. A detailed explanation of this process is in the section "[Dynamic Analysis on Non-Rooted Devices](#dynamic-analysis-on-non-rooted-devices "Dynamic Analysis on Non-Rooted Devices").
```
Comment (commjoen) [inline Document/0x05c-Reverse-Engineering-and-Tampering.md]: This looks a bit weird , do you mean to say: `Wether you are using a rooted phone with frida-server installed or a repackaged app with a Frida gadget on a non-rooted device...`

Comment (commjoen) [inline Document/0x05c-Reverse-Engineering-and-Tampering.md]: Given that you do explain it later, maybe shorten it?
```suggestion
You can dump the app's process memory with [objection](https://github.com/sensepost/objection "Objection") and [Fridump](https://github.com/Nightbringer21/fridump "Fridump"). To take advantage of these tools on a non-rooted device, the Android app must be repackaged with `frida-gadget.so` and re-signed. A detailed explanation of this process is in the section "[Dynamic Analysis on Non-Rooted Devices](#dynamic-analysis-on-non-rooted-devices "Dynamic Analysis on Non-Rooted Devices"). To use these tools on a rooted phone, simply have frida-server installed and running
```

File: Document/0x05c-Reverse-Engineering-and-Tampering.md [hunk 5]
```diff
@@ -1497,7 +1488,368 @@ The hooked function outputted the decrypted string. You extracted the secret str
...
+
+With objection it is possible to dump all memory of the running process on the device by using the command `memory dump all`.
+
+```shell
+$ objection --gadget sg.vantagepoint.helloworldjni explore
+
+sg.vantagepoint.helloworldjni on (google: 8.1.0) [usb] # memory dump all /Users/foo/memory_Android/memory
```
Comment (commjoen) [inline Document/0x05c-Reverse-Engineering-and-Tampering.md]: foo?

File: Document/0x05c-Reverse-Engineering-and-Tampering.md [hunk 6]
```diff
@@ -1497,7 +1488,368 @@ The hooked function outputted the decrypted string. You extracted the secret str
...
+
+Will dump 719 rw- images, totalling 1.6 GiB
+Dumping 1002.8 MiB from base: 0x14140000  [------------------------------------]    0%  00:11:03(session detach message) process-terminated
+Dumping 8.0 MiB from base: 0x7fc753e000  [####################################]  100%
+Memory dumped to file: /Users/foo/memory_Android/memory
+```
+
+In this case there was an error but it could obtain part of the binary. We are able to find the "Hello from C++" strings with radare2:
```
Comment (commjoen) [inline Document/0x05c-Reverse-Engineering-and-Tampering.md]: what was going wrong then? (maybe help the user to find a way to debug it)

File: Document/0x05c-Reverse-Engineering-and-Tampering.md [hunk 7]
```diff
@@ -1497,7 +1488,368 @@ The hooked function outputted the decrypted string. You extracted the secret str
...
+Starting Memory dump...
+Oops, memory access violation!-------------------------------] 0.28% Complete
+Progress: [##################################################] 99.58% Complete
+Running strings on all files:
+Progress: [##################################################] 100.0% Complete
+
+Finished!
+```
+
+It will take a while until it's completed (you might get several memory access violation errors) and you'll get a collection of *.data files inside the dump folder. When you add the `-s` flag, all strings are extracted from the dumped raw memory files and added to the file `strings.txt`, which is also stored in the dump directory.
```
Comment (commjoen) [inline Document/0x05c-Reverse-Engineering-and-Tampering.md]: what do these violation errors mean?

File: Document/0x05c-Reverse-Engineering-and-Tampering.md [hunk 8]
```diff
@@ -1497,7 +1488,368 @@ The hooked function outputted the decrypted string. You extracted the secret str
...
+$ grep -nri owasp-mstg dump/
+Binary file dump//316669952_dump.data matches
+Binary file dump//strings.txt matches
+```
+
+The "owasp-mstg" string can be found in one of the dump files as well as in the processed strings file.
+
+###### Runtime Reverse Engineering
+
+Runtime reverse engineering can be seen as the on-the-fly version of reverse engineering where you don't need, want or can't retrieve the binary data to your host computer. Instead, you'll analyze it straight from the memory of the app.
```
Comment (commjoen) [inline Document/0x05c-Reverse-Engineering-and-Tampering.md]:
```suggestion
Runtime reverse engineering can be seen as the on-the-fly version of reverse engineering where you don't have the binary data to your host computer. Instead, you'll analyze it straight from the memory of the app.
```

File: Document/0x05c-Reverse-Engineering-and-Tampering.md [hunk 9]
```diff
@@ -1497,7 +1488,368 @@ The hooked function outputted the decrypted string. You extracted the secret str
...
+
+Search all symbols of a certain module with `\is <lib>`, e.g. `\is libnative-lib.so`.
+
+```bash
+[0x00000000]> \is libnative-lib.so
+
+[0x00000000]>
+```
+
+Which are empty in this case. Or you might prefer to look into the imports/exports. For example, list the imports with `\ii <lib>`:
```
Comment (commjoen) [inline Document/0x05c-Reverse-Engineering-and-Tampering.md]: Or-->Alternatively

File: Document/0x05c-Reverse-Engineering-and-Tampering.md [hunk 10]
```diff
@@ -1497,7 +1488,368 @@ The hooked function outputted the decrypted string. You extracted the secret str
...
+
+```bash
+[0x00000000]> \ii libnative-lib.so
+0x7dbe1159d0 f __cxa_finalize /system/lib64/libc.so
+0x7dbe115868 f __cxa_atexit /system/lib64/libc.so
+```
+
+And the exports with `\iE <lib>`:
```
Comment (commjoen) [inline Document/0x05c-Reverse-Engineering-and-Tampering.md]:
```suggestion
And list the exports with `\iE <lib>`:
```

File: Document/0x05d-Testing-Data-Storage.md [hunk 11]
```diff
@@ -1049,7 +1049,51 @@ Refer to the `SecureSecretKey` example above for an example `Editable` implement
...
+| /v[1248] value               look for an `cfg.bigendian` 32bit value
+| /w foo                       search for wide string 'f\0o\0o\0'
+| /x ff0033                    search for hex string
+| /z min max                   search for strings of given size
+...
+```
+
+##### Runtime Memory Analysis
+
+Using r2frida you can analyze and inspect the app's memory while running and without needing to dump it. For example, you may run the previous search commands from r2frida and search the memory for a string, hexadecimal values, etc. When doing so, remember to prepend the search command (and any other r2frida specific commands) with a backslash `\` after starting the session with `r2 frida://usb//<name_of_your_app>`.
```
Comment (commjoen) [inline Document/0x05d-Testing-Data-Storage.md]: the first sentence sounds a little bit odd, maybe rephrase it a bit more active?

File: Document/0x06d-Testing-Data-Storage.md [hunk 12]
```diff
@@ -802,112 +802,49 @@ In summary, when performing static analysis for sensitive data exposed via memor
 
 #### Dynamic Analysis
 
-Several approaches and tools are available for dumping an iOS app's memory.
+Several approaches and tools are available for testing the memory of an iOS app.
```
Comment (commjoen) [inline Document/0x06d-Testing-Data-Storage.md]: so are we testing the memory? or are we searching it? or testing it for the presence of sensitive information? or... dumping alone is a bit too little indeed, but what are we testing it for :) ?

File: Document/0x06d-Testing-Data-Storage.md [hunk 13]
```diff
@@ -802,112 +802,49 @@ In summary, when performing static analysis for sensitive data exposed via memor
...
 
-On a non-jailbroken device, you can dump the app's process memory with [objection](https://github.com/sensepost/objection "Objection") and [Fridump](https://github.com/Nightbringer21/fridump "Fridump"). To take advantage of these tools, the iOS app must be repackaged with `FridaGadget.dylib` and re-signed. A detailed explanation of this process is in the section "Dynamic Analysis on Non-Jailbroken Devices", in the chapter "Tampering and Reverse Engineering on iOS".
+##### Retrieving and Analyzing a Memory Dump
 
-##### Objection (No Jailbreak needed)
+Wether you are using a jailbroken or a non-jailbroken device, you can dump the app's process memory with [objection](https://github.com/sensepost/objection "Objection") and [Fridump](https://github.com/Nightbringer21/fridump "Fridump"). Find a detailed explanation of this process in the section "[Memory Dump](0x06c-Reverse-Engineering-and-Tampering.md#memory-dump "Memory Dump")", in the chapter "Tampering and Reverse Engineering on iOS".
```
Comment (commjoen) [inline Document/0x06d-Testing-Data-Storage.md]: Find a detailed  --> You can find a

File: Document/0x06c-Reverse-Engineering-and-Tampering.md [hunk 14]
```diff
@@ -609,6 +609,379 @@ Start Safari on the iOS device. Run the above Python script on your connected ho
...
 
+##### Process Exploration
+
+When testing an app, process exploration can provide the tester with deep insights into the app process memory. It can be achieved via runtime instrumentation and allows to perform tasks such as:
+
+- Retrieving the memory map and loaded libraries.
+- Searching for occurrences of certain data.
+- After doing a search, obtaining the location of a certain offset in the memory map.
+- Performing a memory dump and inspect or reverse engineer the binary data _offline_.
+- Reverse engineering a binary or Framework while it's running.
```
Comment (commjoen) [inline Document/0x06c-Reverse-Engineering-and-Tampering.md]: this seems to be a bit more repetitive.. maybe some of the general testing can be centralized later? (not now...)

File: Document/0x06c-Reverse-Engineering-and-Tampering.md [hunk 15]
```diff
@@ -609,6 +609,379 @@ Start Safari on the iOS device. Run the above Python script on your connected ho
...
+
+As you can see, these tasks are rather supportive and/or passive, they'll help us collect data and information that will support other techniques. Therefore, they're normally used in combination with other techniques such as method hooking.
```
Comment (commjoen) [inline Document/0x06c-Reverse-Engineering-and-Tampering.md]: see earlier comment 👍

File: Document/0x06c-Reverse-Engineering-and-Tampering.md [hunk 16]
```diff
@@ -609,6 +609,379 @@ Start Safari on the iOS device. Run the above Python script on your connected ho
...
+...
+0x0000000100ea8000 - 0x0000000100eb0000 rw-
+0x0000000100eb0000 - 0x0000000100eb4000 r--
+0x0000000100eb4000 - 0x0000000100eb8000 r-x /usr/lib/TweakInject.dylib
+0x0000000100eb8000 - 0x0000000100ebc000 rw- /usr/lib/TweakInject.dylib
+0x0000000100ebc000 - 0x0000000100ec0000 r-- /usr/lib/TweakInject.dylib
+0x0000000100f60000 - 0x00000001012dc000 r-x /private/var/containers/Bundle/Application/3ADAF47D-A734-49FA-B274-FBCA66589E67/iGoat-Swift.app/Frameworks/Realm.framework/Realm
+```
+
+While you're searching or exploring the app memory, you can always verify where you're located in each moment (where your current offset is located) in the memory map. Instead of noting and searching for the memory address in this list you can simply run `\dm.`. You'll find an example in the following section "In-Memory Search".
```
Comment (commjoen) [inline Document/0x06c-Reverse-Engineering-and-Tampering.md]:
```suggestion
While you're searching or exploring the app memory, you can always verify where your current offset is located in the memory map. Instead of noting and searching for the memory address in this list you can simply run `\dm.`. You'll find an example in the following section "In-Memory Search".
```

File: Document/0x06c-Reverse-Engineering-and-Tampering.md [hunk 17]
```diff
@@ -609,6 +609,379 @@ Start Safari on the iOS device. Run the above Python script on your connected ho
...
+0x000000018ea1d000 UIKit
+0x0000000100f60000 Realm
+```
+
+As you might expect you can correlate the addresses of the libraries with the memory maps: e.g. the main app binary iGoat-Swift is located at `0x0000000100b7c000` and the Realm Framework at `0x0000000100f60000`.
+
+You can also use objection to display the same information.
+
+```shell
+OWASP.iGoat-Swift on (iPhone: 11.1.2) [usb] # memory list modules
```
Comment (commjoen) [inline Document/0x06c-Reverse-Engineering-and-Tampering.md]: don't forget to guide the user on how to get there.

File: Document/0x06c-Reverse-Engineering-and-Tampering.md [hunk 18]
```diff
@@ -609,6 +609,379 @@ Start Safari on the iOS device. Run the above Python script on your connected ho
...
+
+```bash
+[0x00000000]> s 0x100d7d332
+[0x100d7d332]> \dm.
+0x0000000100b7c000 - 0x0000000100de0000 r-x /private/var/containers/Bundle/Application/3ADAF47D-A734-49FA-B274-FBCA66589E67/iGoat-Swift.app/iGoat-Swift
+```
+
+As expected, you are located in the region of the main iGoat-Swift binary (r-x, read and execute). In the previous section, you saw that the main binary is located between `0x0000000100b7c000` and `0x0000000100e97000`.
+
+Now, for this second example, you can search for something that's not in the app binary nor in any loaded library, typically user input. Open the iGoat-Swift app and navigate in the menu to Authentication -> Remote Authentication -> Start. There you'll find a password field that you can overwrite. Write the string "owasp-mstg" but do not click on Login just yet. You will perform two steps.
```
Comment (commjoen) [inline Document/0x06c-Reverse-Engineering-and-Tampering.md]: You will perform two steps. --> Instead perform the following two steps.

File: Document/0x06c-Reverse-Engineering-and-Tampering.md [hunk 19]
```diff
@@ -609,6 +609,379 @@ Start Safari on the iOS device. Run the above Python script on your connected ho
...
+
+```bash
+[0x00000000]> \/ owasp-mstg
+hits: 1
+0x1c06619c0 hit3_0 owasp-mstg
+
+[0x100d7d332]> s 0x1c06619c0
+[0x1c06619c0]> \dm.
+0x00000001c0000000 - 0x00000001c8000000 rw-
```
Comment (commjoen) [inline Document/0x06c-Reverse-Engineering-and-Tampering.md]: these steps are not really clear to me...

File: Document/0x06c-Reverse-Engineering-and-Tampering.md [hunk 20]
```diff
@@ -609,6 +609,379 @@ Start Safari on the iOS device. Run the above Python script on your connected ho
...
+
+In-memory search can be very useful to quickly know if certain data is located in the main app binary, inside a shared library or in another region. You may also use it to test the behavior of the app regarding how the data is kept in memory. For instance, you could continue the previous example, this time clicking on Login and searching again for occurrences of the data. Also, you may check if you still can find those strings in memory after the login is completed to verify if this _sensitive data_ is wiped from memory after its use.
+
+###### Memory Dump
+
+Wether you are using a jailbroken device with frida-server installed or a non-jailbroken device, you can dump the app's process memory with [objection](https://github.com/sensepost/objection "Objection") and [Fridump](https://github.com/Nightbringer21/fridump "Fridump"). To take advantage of these tools on a non-jailbroken device, the iOS app must be repackaged with `FridaGadget.dylib` and re-signed. A detailed explanation of this process is in the section "[Dynamic Analysis on Non-Jailbroken Devices](#dynamic-analysis-on-non-jailbroken-devices "Dynamic Analysis on Non-Jailbroken Devices").
```
Comment (commjoen) [inline Document/0x06c-Reverse-Engineering-and-Tampering.md]: see earlier comments.

File: Document/0x06c-Reverse-Engineering-and-Tampering.md [hunk 21]
```diff
@@ -609,6 +609,379 @@ Start Safari on the iOS device. Run the above Python script on your connected ho
...
+iPhone on (iPhone: 10.3.1) [usb] # memory dump all /Users/foo/memory_iOS/memory
+Dumping 768.0 KiB from base: 0x1ad200000  [####################################]  100%
+Memory dumped to file: /Users/foo/memory_iOS/memory
+```
+
+Alternatively you can use Fridump.
+
+> The original version of Fridump is no longer maintained, and the tool works only with Python 2. The latest Python version (3.x) should be used for Frida, so Fridump doesn't work out of the box.
+>
+> If you're getting the following error message despite your iOS device being connected via USB, checkout [Fridump with the fix for Python 3](https://github.com/sushi2k/fridump "Fridump for Python3").
```
Comment (commjoen) [inline Document/0x06c-Reverse-Engineering-and-Tampering.md]: so how can we make this work?

File: Document/0x06c-Reverse-Engineering-and-Tampering.md [hunk 22]
```diff
@@ -609,6 +609,379 @@ Start Safari on the iOS device. Run the above Python script on your connected ho
...
+Note that in order to find this string using the `strings` command you'll have to specify an encoding using the `-e` flag and in this case `l` for 16-bit little-endian character.
+
+```bash
+$ strings -e l memory_ios | grep owasp-mstg
+owasp-mstg
+```
+
+###### Runtime Reverse Engineering
+
+Runtime reverse engineering can be seen as the on-the-fly version of reverse engineering where you don't need, want or can't retrieve the binary data to your host computer. Instead, you'll analyze it straight from the memory of the app.
```
Comment (commjoen) [inline Document/0x06c-Reverse-Engineering-and-Tampering.md]: see earlier comments.

File: Document/0x05c-Reverse-Engineering-and-Tampering.md (hunk 3 above)
Comment (cpholguera) [reply Document/0x05c-Reverse-Engineering-and-Tampering.md]: nice point, I'll add something on that ;)

**Comment by OMTGreleaser** [comment]
Results for commit 0681497c607a69c044633b0a8a048fd8f09fae4a: Broken link result: [!] Error(s) found by the Links validator: 6 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

File: Document/0x06c-Reverse-Engineering-and-Tampering.md (hunk 21 above)
Comment (cpholguera) [reply Document/0x06c-Reverse-Engineering-and-Tampering.md]: Actually I didn't have any problems. It just worked following the standard instructions. I'd remove these paragraphs (someone put them there before me).

File: Document/0x06c-Reverse-Engineering-and-Tampering.md (hunk 14 above)
Comment (cpholguera) [reply Document/0x06c-Reverse-Engineering-and-Tampering.md]: Totally agree :)

**Comment by OMTGreleaser** [comment]
Results for commit 7a74d058dff9ede73ca31c3acf56ead88e994643: Broken link result: [!] Error(s) found by the Links validator: 7 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

**Comment by OMTGreleaser** [comment]
Results for commit e0160c66426f07a65f2a5acb165f2663340e1add: Broken link result: [!] Error(s) found by the Links validator: 6 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

**Comment by OMTGreleaser** [comment]
Results for commit 89350e681d01c501fad3e40b2eaa478c14f82d75: Broken link result: [!] Error(s) found by the Links validator: 6 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

**Comment by OMTGreleaser** [comment]
Results for commit f458730913e35e9ef4baca4082c6b5e1b896ca43: Broken link result: [!] Error(s) found by the Links validator: 5 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

File: Document/0x05c-Reverse-Engineering-and-Tampering.md (hunk 1 above)
Comment (cpholguera) [reply Document/0x05c-Reverse-Engineering-and-Tampering.md]: done

File: Document/0x05c-Reverse-Engineering-and-Tampering.md (hunk 2 above)
Comment (cpholguera) [reply Document/0x05c-Reverse-Engineering-and-Tampering.md]: done

File: Document/0x05c-Reverse-Engineering-and-Tampering.md (hunk 3 above)
Comment (cpholguera) [reply Document/0x05c-Reverse-Engineering-and-Tampering.md]: done

File: Document/0x05c-Reverse-Engineering-and-Tampering.md (hunk 4 above)
Comment (cpholguera) [reply Document/0x05c-Reverse-Engineering-and-Tampering.md]: done

Comment (cpholguera) [reply Document/0x05c-Reverse-Engineering-and-Tampering.md]: done

File: Document/0x05c-Reverse-Engineering-and-Tampering.md (hunk 9 above)
Comment (cpholguera) [reply Document/0x05c-Reverse-Engineering-and-Tampering.md]: done

File: Document/0x05d-Testing-Data-Storage.md (hunk 11 above)
Comment (cpholguera) [reply Document/0x05d-Testing-Data-Storage.md]: done

File: Document/0x06d-Testing-Data-Storage.md (hunk 12 above)
Comment (cpholguera) [reply Document/0x06d-Testing-Data-Storage.md]: done

File: Document/0x06d-Testing-Data-Storage.md (hunk 13 above)
Comment (cpholguera) [reply Document/0x06d-Testing-Data-Storage.md]: done

File: Document/0x06c-Reverse-Engineering-and-Tampering.md (hunk 17 above)
Comment (cpholguera) [reply Document/0x06c-Reverse-Engineering-and-Tampering.md]: done

File: Document/0x06c-Reverse-Engineering-and-Tampering.md (hunk 18 above)
Comment (cpholguera) [reply Document/0x06c-Reverse-Engineering-and-Tampering.md]: done

File: Document/0x06c-Reverse-Engineering-and-Tampering.md (hunk 20 above)
Comment (cpholguera) [reply Document/0x06c-Reverse-Engineering-and-Tampering.md]: done

File: Document/0x06c-Reverse-Engineering-and-Tampering.md (hunk 22 above)
Comment (cpholguera) [reply Document/0x06c-Reverse-Engineering-and-Tampering.md]: done

**Comment by OMTGreleaser** [comment]
Results for commit a44211d2824cc05309f913bd8d536eb44a9e5ade: Broken link result: [!] Error(s) found by the Links validator: 6 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

File: Document/0x06c-Reverse-Engineering-and-Tampering.md (hunk 19 above)
Comment (cpholguera) [reply Document/0x06c-Reverse-Engineering-and-Tampering.md]: done

File: Document/0x05c-Reverse-Engineering-and-Tampering.md (hunk 6 above)
Comment (cpholguera) [reply Document/0x05c-Reverse-Engineering-and-Tampering.md]: both this and the next are related. I've added a comment until I have more concrete details or explanation. For now it is safe to ignore and the user doesn't have to debug it or fix it. It's an issue in the Frida agent trying to read where it's not allowed to.

File: Document/0x05c-Reverse-Engineering-and-Tampering.md (hunk 7 above)
Comment (cpholguera) [reply Document/0x05c-Reverse-Engineering-and-Tampering.md]: both this and the previous are related. I've added a comment until I have more concrete details or explanation. For now it is safe to ignore and the user doesn't have to debug it or fix it. It's an issue in the Frida agent trying to read where it's not allowed to.

**Comment by OMTGreleaser** [comment]
Results for commit 1035df76b043b80ea9ec037392816a221d73cb1f: Broken link result: [!] Error(s) found by the Links validator: 5 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

**Comment by cpholguera** [comment]
This last commit addresses the memory access violations cause, which is the tools making a brute-dump of all addresses in the memory map. Some of them are in fact not readable, which provoked those violations. Added more contextual information.

File: Document/0x05c-Reverse-Engineering-and-Tampering.md [hunk 23]
```diff
@@ -1497,7 +1488,378 @@ The hooked function outputted the decrypted string. You extracted the secret str
...
+
+When testing an app, process exploration can provide the tester with deep insights into the app process memory. It can be achieved via runtime instrumentation and allows to perform tasks such as:
+
+- Retrieving the memory map and loaded libraries.
+- Searching for occurrences of certain data.
+- After doing a search, obtaining the location of a certain offset in the memory map.
+- Performing a memory dump and inspect or reverse engineer the binary data _offline_.
+- Reverse engineering a native library while it's running.
+
+As you can see, these passive tasks help us collect information. Information is often used for other techniques, such as method hooking.
```
Comment (commjoen) [inline Document/0x05c-Reverse-Engineering-and-Tampering.md]:
```suggestion
As you can see, these passive tasks help us collect information. This Information is often used for other techniques, such as method hooking.
```

File: Document/0x05c-Reverse-Engineering-and-Tampering.md [hunk 24]
```diff
@@ -1497,7 +1488,378 @@ The hooked function outputted the decrypted string. You extracted the secret str
...
+0x0000007dc065d000 linux-vdso.so.1
+0x0000007dc065f000 linker64
+```
+
+As you might expect you can correlate the addresses of the libraries with the memory maps: e.g. the native library of the app is located at `0x0000007d1c499000` and optimized dex (base.odex) at `0x0000007d10dd0000`.
+
+You can also use objection to display the same information.
+
+```shell
+$ objection --gadget sg.vantagepoint.helloworldjni explore
```
Comment (commjoen) [inline Document/0x05c-Reverse-Engineering-and-Tampering.md]: maybe make sure you explain somewhere how to get the gadget in... or refer to that chapter where we explain how to use objection earlier?

File: Document/0x06c-Reverse-Engineering-and-Tampering.md [hunk 25]
```diff
@@ -609,6 +609,373 @@ Start Safari on the iOS device. Run the above Python script on your connected ho
...
+0x000000018ea1d000 UIKit
+0x0000000100f60000 Realm
+```
+
+As you might expect you can correlate the addresses of the libraries with the memory maps: e.g. the main app binary iGoat-Swift is located at `0x0000000100b7c000` and the Realm Framework at `0x0000000100f60000`.
+
+You can also use objection to display the same information.
+
+```shell
+$ objection --gadget OWASP.iGoat-Swift explore
```
Comment (commjoen) [inline Document/0x06c-Reverse-Engineering-and-Tampering.md]: dont' forget to guide the user on how to use objection; or is that chapter close by?

File: Document/0x05c-Reverse-Engineering-and-Tampering.md (hunk 24 above)
Comment (commjoen) [reply Document/0x05c-Reverse-Engineering-and-Tampering.md]: ah it is there, but way earlier. sorry!
